src/platforms/hosted/bmp_libusb.c

```diff
@@ -47,291 +106,373 @@ void libusb_exit_function(bmp_info_s *info)
-					if (strncmp(cable->description, product, strlen(cable->description)) != 0)
-						continue; /* descriptions do not match*/
-					found = true;
-				} else {                                 /* VID/PID fits, but no cl_opts->opt_cable and no description*/
-					if (cable->vendor == 0x0403 &&       /* FTDI*/
-						(cable->product == 0x6010 ||     /* FT2232C/D/H*/
-							cable->product == 0x6011 ||  /* FT4232H Quad HS USB-UART/FIFO IC */
-							cable->product == 0x6014)) { /* FT232H Single HS USB-UART/FIFO IC */
-						ftdi_unknown = true;
-						continue; /* Cable name is needed */
-					}
+				//
+				// Now read any strings that have not been set by a custom reader function
+				//
+				if (product == NULL) {
+					product = get_device_descriptor_string(handle, device_descriptor->iProduct);
 				}
-				if (found) {
-					active_cable = cable->name;
-					type = BMP_TYPE_FTDI;
-					break;
+				if (manufacturer == NULL) {
+					manufacturer = get_device_descriptor_string(handle, device_descriptor->iManufacturer);
+				}
+				if (serial == NULL) {
+					//
+					// If this is an STLink V2 it does not report serial number correctly, for bypass
+					// trying to read it
+					//
+					if (device_descriptor->idVendor != VENDOR_ID_STLINK &&
+						device_descriptor->idProduct != PRODUCT_ID_STLINKV2)
+						serial = get_device_descriptor_string(handle, device_descriptor->iSerialNumber);
+					else
+						serial = strdup("---");
+				}
+
+				if (version == NULL) {
+					version = strdup("---");
 				}
+
+				*probe_list = probe_info_add_by_id(*probe_list, probe_type, device, device_descriptor->idVendor,
+					device_descriptor->idProduct, manufacturer, product, serial, version);
+				probe_added = true;
+				libusb_close(handle);
 			}
-			if (!cable->name)
-				continue;
+			break;
 		}
+		vid_pid_index++;
+	}
+	return probe_added;
+}
 
-		if (report)
-			DEBUG_WARN("%2zu: %s, %s, %s\n", found_debuggers + 1U, serial[0] ? serial : NO_SERIAL_NUMBER, manufacturer,
-				product);
-
-		info->libusb_dev = dev;
-		info->vid = desc.idVendor;
-		info->pid = desc.idProduct;
-		info->bmp_type = type;
-		strncpy(info->serial, serial, sizeof(info->serial));
-		strncpy(info->product, product, sizeof(info->product));
-		strncpy(info->manufacturer, manufacturer, sizeof(info->manufacturer));
-		if (cl_opts->opt_position && cl_opts->opt_position == found_debuggers + 1U) {
-			found_debuggers = 1;
-			break;
+static const probe_info_s *scan_for_devices(bmp_info_s *info)
+{
+	libusb_device **device_list;
+	libusb_device_descriptor_s device_descriptor;
+	probe_info_s *probe_list = NULL;
+
+	int result;
+	ssize_t cnt;
+	//
+	// If we are running on Windows the proprietary FTD2XX library is used
+	// to collect debugger information.
+	//
+#if defined(_WIN32) || defined(__CYGWIN__)
+	probe_list = process_ftdi_probe();
+	const bool skip_ftdi = probe_list != NULL;
+#else
+	const bool skip_ftdi = false;
+#endif
+
+	cnt = libusb_get_device_list(NULL, &device_list);
```

Comment:
```suggestion
	cnt = libusb_get_device_list(info->libusb_ctx, &device_list);
```
Using this with NULL for default context (straight from libusb doc examples) segfaults inside libusb-1.0.25 on ubuntu-jammy.
Otherwise, with your PR, BMDA works with "J-Link STLink V21" OB on Nucleo-64, even with RTT (at only 4500 chars/s).

Reply:
Elaborating on this: apparently libusb-1.0.26 (which I have tested on my Gentoo PC) incorporates a workaround. The expected API usage looks like to either do `libusb_init(NULL)` (BMDA doesn't) to allocate a default context, or to use the new allocated context everywhere without reverting to NULLs. 
I see this output for this branch but no crashes:
```
altracer@glacier ~/g/s/blackmagic (feature/improved-bmda-probe-scan)> env LIBUSB_DEBUG=3 ./src/blackmagic -tv 1
libusb: warning [libusb_init] installing new context as implicit default
libusb: error [usbi_get_context] API misuse! Using non-default context as implicit default.
No probes found
```

Reply:
EDit as suggested